File: macos/Classes/src/ThumbnailGenerator.swift
import Foundation
import ffmpegkit

class ThumbnailGenerator {

    static func generateThumbnails(
        videoBytes: Data,
        timestamps: [Double],
        thumbnailFormat: String,
        extension ext: String,
        width: Int,
        completion: @escaping ([Data]) -> Void
    ) {
        DispatchQueue.global(qos: .userInitiated).async {
            let tempVideoURL = writeBytesToTempFile(videoBytes, ext: ext)
            var thumbnails: [Data?] = Array(repeating: nil, count: timestamps.count)
            let group = DispatchGroup()

            for (index, time) in timestamps.enumerated() {
                group.enter()
                let timestampStr = String(format: "%.3f", time / 1000.0)
                let imageFileURL = FileManager.default.temporaryDirectory
                    .appendingPathComponent("thumb_\(index).\(thumbnailFormat)")

                let command =
                    "-y -ss \(timestampStr) -i \(tempVideoURL.path) -vframes 1 -vf scale=\(width):-2 \(imageFileURL.path)"

                let startTime = Date().timeIntervalSince1970

                FFmpegKit.executeAsync(command) { session in
                    let duration = Int((Date().timeIntervalSince1970 - startTime) * 1000)
                    if let returnCode = session?.getReturnCode(), ReturnCode.isSuccess(returnCode) {
                        if let imageData = try? Data(contentsOf: imageFileURL) {
                            thumbnails[index] = imageData
                            print("[\(index)] ✅ \(timestampStr)s in \(duration)ms (\(imageData.count) bytes)")
                        }
                    } else {
                        print("[\(index)] ❌ Failed at \(timestampStr)s in \(duration)ms")
                    }

                    try? FileManager.default.removeItem(at: imageFileURL)
                    group.leave()
                }
            }

            group.wait()
            try? FileManager.default.removeItem(at: tempVideoURL)

            DispatchQueue.main.async {
                completion(thumbnails.compactMap { $0 })
            }
        }
    }

    private static func writeBytesToTempFile(_ bytes: Data, ext: String) -> URL {
        let fileURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("video_temp.\(ext)")
        try? bytes.write(to: fileURL)
        return fileURL
    }
}

File: android/src/main/kotlin/ch/waio/pro_video_editor/src/ThumbnailGenerator.kt
// =======================
// ThumbnailGenerator.kt
// =======================

package ch.waio.pro_video_editor.src

import android.content.Context
import android.media.MediaMetadataRetriever
import android.util.Log
import com.arthenica.ffmpegkit.FFmpegKit
import com.arthenica.ffmpegkit.ReturnCode
import kotlinx.coroutines.*
import java.io.File
import java.io.FileOutputStream

class ThumbnailGenerator(private val context: Context) {

    suspend fun generateThumbnails(
        videoBytes: ByteArray,
        timestamps: List<Long>,
        thumbnailFormat: String,
        extension: String,
        width: Int
    ): List<ByteArray> = withContext(Dispatchers.IO) {
        val TAG = "FFmpegThumbnailGen"
        val tempVideoFile = writeBytesToTempFile(videoBytes, extension)
        val thumbnails = MutableList<ByteArray?>(timestamps.size) { null }

        val jobs = timestamps.mapIndexed { index, timeMs ->
            async {
                val startTime = System.currentTimeMillis()
                val tempImageFile =
                    File.createTempFile("thumb_$index", ".$thumbnailFormat", context.cacheDir)
                val timestampStr = String.format("%.3f", timeMs / 1000.0)

                val command =
                    "-y -ss $timestampStr -i ${tempVideoFile.absolutePath} -vframes 1 -vf scale=$width:-2 ${tempImageFile.absolutePath}"
                val session = FFmpegKit.execute(command)
                val duration = System.currentTimeMillis() - startTime

                if (ReturnCode.isSuccess(session.returnCode)) {
                    val bytes = tempImageFile.readBytes()
                    thumbnails[index] = bytes
                    Log.d(TAG, "[$index] ✅ $timestampStr s in $duration ms (${bytes.size} bytes)")
                } else {
                    Log.w(TAG, "[$index] ❌ Failed at $timestampStr s in $duration ms")
                }

                tempImageFile.delete()
            }
        }

        jobs.awaitAll()
        tempVideoFile.delete()
        return@withContext thumbnails.filterNotNull()
    }


    private fun writeBytesToTempFile(bytes: ByteArray, extension: String): File {
        val tempFile = File.createTempFile("video_temp", ".$extension", context.cacheDir)
        FileOutputStream(tempFile).use {
            it.write(bytes)
        }
        return tempFile
    }
}
